ReactNativeClient/lib/services/KeymapService.ts

```diff
@@ -0,0 +1,202 @@
+			for (let i = 0; i < customKeymap.length; i++) {
+				const item = customKeymap[i];
+
+				// Validation errors will be caught and logged
+				if (!item.hasOwnProperty('command')) {
+					this.logger().warn(`Command is missing in: ${JSON.stringify(item)}`);
+				} else if (!item.hasOwnProperty('accelerator')) {
+					this.logger().warn(`Accelerator is missing in: ${JSON.stringify(item)}`);
+				} else if (!this.keymap.hasOwnProperty(item.command)) {
+					this.logger().warn(`Command is invalid in: ${JSON.stringify(item)}`);
+				} else if (!(item.accelerator === null || KeymapService.isAccelerator(item.accelerator))) {
+					this.logger().warn(`Accelerator is invalid in: ${JSON.stringify(item)}`);
+				} else {
+					this.setAccelerator(item.command, item.accelerator);
+				}
+			}
+
+			const currentPlatform = shim.platformName();
+			const usedAccelerators = new Set();
+
+			// Detect duplicate Accelerators in the keymap
+			for (const item of Object.values(this.keymap)) {
+				const itemPlatforms = new Set(item.platforms);
+
+				// Only validate against the current platform
+				// Different platforms may have the same Accelerator for different commands
+				// i.e. Accelerators for Options and Preferences are the same value: "CommandOrControl+,"
+
+				if (itemPlatforms.has(currentPlatform)) {
+					if (usedAccelerators.has(item.accelerator)) {
+						// Should reset the keymap to default values
+						this.resetKeymap();
+
+						this.logger().warn(`Accelerator is duplicate in: ${JSON.stringify({ command: item.command, accelerator: item.accelerator })}`);
+						return reject(new Error('Falling back to default keymap!'));
+					}
+
+					// There can be multiple "null" Accelerators
+					if (item.accelerator !== null) {
+						usedAccelerators.add(item.accelerator);
+					}
+				}
+			}
+
+			resolve(); // All good!
+		});
+	}
+
+	resetKeymap() {
+		this.keymap = KeymapService.initKeymap();
+	}
+
+	static initKeymap() {
+		const itemsByCommand: { [command: string]: KeymapItem } = {};
+		for (let i = 0; i < KeymapService.defaultKeymap.length; i++) {
+			// We should make a copy of the KeymapItem before assigning it, hence the spread operator
+			// Otherwise we're going to mess up the defaultKeymap array
+			itemsByCommand[KeymapService.defaultKeymap[i].command] = { ...KeymapService.defaultKeymap[i] };
+		}
+
+		return itemsByCommand;
+	}
+
+	static isAccelerator(accelerator: string) {
+		let keyFound = false;
+		const parts = accelerator.split('+');
+
+		return parts.every((part, index) => {
+			const isKey = KeymapService.keyCodes.test(part);
+			const isModifier = KeymapService.modifiers.test(part);
+
+			if (isKey) {
+				// Key must be unique
+				if (keyFound) return false;
+				keyFound = true;
+			}
+
+			// Key is required
+			if (index === (parts.length - 1) && !keyFound) return false;
+			return isKey || isModifier;
+		});
+	}
+
+	static localizeAccelerator(_accelerator: string, reverse = false) {
```

Comment:
Why an underscore in `_accelerator`?

Reply:
What does this function do?

Reply:
This is a utility function for the GUI editor. Basically it converts "CommandOrControl+X" into "Control+X" on Windows/Linux and "Command+X" on Mac. If we're seperating the default keymaps then this is no longer needed.